### DISCUSSION TITLE
SERVICE UNAVAILABLE error for Oauth2 with both local instance of keycloak and local instance of custom SSO

### DISCUSSION THREAD

**Discussion by William-Kara** [directus/directus#13805] [Archive]
### Describe the Bug

I am trying to make a custom SSO work with directus, I am also trying with a Keycloak to make sure the issue isn't coming from the SSO.
After login on the SSO, I keep getting SERVICE_UNAVAILABLE error. It works fine with online SSO platforms like github/facebook/google.
It is pretty similare to https://github.com/directus/directus/issues/7427 and https://github.com/directus/directus/issues/6749 but in the later case I checked and this doesn't seems to be the issue, and the first one I don't realy get how to translate that when I'm not in a docker setup.

### To Reproduce

Start a keycloak following the Docker guide https://www.keycloak.org/getting-started/getting-started-docker
Deploy a directus with npx.

Configure keycloak realm and client and add the configuration of keycloak in the directus .env like so:

```
## General

HOST="localhost"
PORT=8055
PUBLIC_URL="http://localhost:8055"
```
```
## Auth Providers
AUTH_PROVIDERS="keycloak"

AUTH_KEYCLOAK_DRIVER="openid"
AUTH_KEYCLOAK_CLIENT_ID="directus-sso"
AUTH_KEYCLOAK_CLIENT_SECRET="0sffjbgd1Ukg1jU8SCD0fyw6HiJrrwQIPOPOP"
AUTH_KEYCLOAK_ISSUER_URL="http://localhost:8080/realms/myrealm/.well-known/openid-configuration"
AUTH_KEYCLOAK_IDENTIFIER_KEY="email"
AUTH_KEYCLOAK_SCOPE="email"
AUTH_KEYCLOAK_ALLOW_PUBLIC_REGISTRATION="true"
AUTH_KEYCLOAK_DEFAULT_ROLE_ID="0a1d1798-c79f-4c54-aa12-6c46802ff109"
```

### Errors Shown

http://localhost:8055/admin/login?reason=SERVICE_UNAVAILABLE

### What version of Directus are you using?

9.11.1

### What version of Node.js are you using?

v14.19.0

### What database are you using?

Posgres 14

### What browser are you using?

Firefox, chrome

### How are you deploying Directus?

running locally

**Comment by aidenfoxx** [comment]
Not sure this will be the issue, but to use OpenID correctly you should have it in your scope `AUTH_KEYCLOAK_SCOPE="openid email"`.

Otherwise I would advise setting your Directus logging level to "trace" (and maybe set style to "raw") to get a better idea of what's going wrong in the server logs.

**Comment by William-Kara** [comment]
Thank you, the trace logging level with style raw allowed me to see my issue for the keycloak openid configuration.

On the other hand, do you know if they is a way to get a log of the request sent to the AUTH_<PROVIDER>_ACCESS_URL ?

**Comment by aidenfoxx** [comment]
The "ACCESS_URL" variable shouldn't be used with OpenID. It's only applicable to OAuth 2.0 integrations.

It is currently not possible to see the network requests made as part of SSO. We use a module called `openid-client` which handles the network stuff, so we are unable to see what requests are being sent.

**Comment by William-Kara** [comment]
Like I was saying in my first post, I am trying to work with another custom sso that is using Oauth2.
It is in this case that I am trying to see the request, because I get a "missing field `client_id`" error from it on my "ACCESS_URL".
It my be me not understanding completly the flow, but this must be the API call for the token right ? So how can it not have the client_id ?

**Reply by aidenfoxx** [reply]
Yes, that would be the token URL. I'm not sure what `openid-client` sends to the URL. The code for handling token grants is here but it's kinda hard to follow: https://github.com/panva/node-openid-client/blob/3f8dde8b891382d9adac586255c521f444b4bb81/lib/client.js#L1310

**Comment by William-Kara** [comment]
Hello, 
I managed to see the request and it seems to indeed lack the parameters client_id and client_secret that are required on the SSO side.
<img width="1249" alt="Capture d’écran 2022-06-23 à 10 55 16" src="https://user-images.githubusercontent.com/8873780/175262058-a532d954-c60d-4e4d-802b-8822d9053215.png">

I would ask for those parameter to not be required but every documentation I can find about the authorization_code grant_type have both of them in the list of parameters.

Do you have any idea how to solve this ?

**Reply by aidenfoxx** [reply]
The secret and client id is sent as an encoded authorization header, which is why you don't see them in the form data.

I know it's not much help to you, but Keycloak has been tested as working by multiple people on our side. I would advise making sure your configuration is correct on the SSO side, or seeking more technical assistance from the OpenID library repo: https://github.com/panva/node-openid-client
